README.md
# expat-fatcat

`expat-fatcat` helps US taxpayers living abroad file their tax returns to comply with the IRS and [FATCA](https://en.wikipedia.org/wiki/Foreign_Account_Tax_Compliance_Act). A significant pain point in this process is that foreign payments (income or deductions) must sometimes be converted to USD with a valid rate on the date of payment. This can readily add up to 40+ historical FX-rates to look up and then paste into the usual Excel accounting madness. Some people may consider a little copy-paste drudge work once a year a minor annoyance, but what are programming and APIs for if not injecting some fun into an otherwise dreary task?

The core functionality of `expat-fatcat` is an historical FX-rate lookup with smoothing in case of a missing exchange rate. Currently, we use the FX-service [Quandl](https://www.quandl.com/) in `QuandlRateCoverterToUSD`, which is a sub-class of ``AbsRateConverterTo``. Other FX-rate services could be readily integrated as required.

```mermaid
    flowchart LR
        start(( ))
        login([user logs in or\n starts console / notebook])
        enter([user enters foreign\npayments with currency\nand date])
        aggregate([user requests sums in USD for manual entry])
        output([user requests file output\nof conversions\nas supporting material])
        start --> login --> enter --> aggregate
        enter --> output
```

## Installation

Install from [PyPI](https://pypi.org/) with `pip install expat-fatcat`.

## Example usage

```python
from expat_fatcat import QuandlRateConverterToUSD, FatcatCalculator
converter = QuandlRateConverterToUSD()
converter.get_rate('EUR', datetime(2019, 4, 18))
# 1.125
payments = [dict(date='2018-01-15', amount=100), dict(date='2018-02-15', amount=150)]
calculator = FatcatCalculator(converter)
calculator(payments, 'EUR', '%Y-%m-%d')
# 310.165
```

See also [notebooks/example-usage.ipynb](notebooks/example-usage.ipynb).

To make more than 50 API calls / day, save your Quandl API key in the environment variable `QUANDL_KEY`, and it will be used upon initialization of `QuandlRateConverterToUSD`.


release-notes.md
# Release notes

## Latest

* [feature] Add release notes
* [doc] Add user story flowchart to readme
* [doc] Fix typos in README

## v0.2.5 and before

Changes made in the times before release notes :smile:


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for expat_fatcat."""


tests/test_expat_fatcat.py
"""Tests for `expat_fatcat` package."""
import pytest
import numpy as np
from datetime import datetime

import pandas as pd

import expat_fatcat
from expat_fatcat.expat_fatcat import DummyRateConverterTo


class TestConverter:

    def test_asymmetric_rate_smoothing(self):
        dates_fx_rates = [
            (datetime(2017, 2, 19), 2.2),
            (datetime(2017, 2, 20), np.nan),
            (datetime(2017, 2, 21), np.nan),
        ]
        converter = DummyRateConverterTo('FOO', dates_fx_rates)
        assert converter.get_rate('FOO', datetime(2017, 2, 19)) == 2.2
        assert converter.get_rate('FOO', datetime(2017, 2, 20)) == 2.2

    def test_symmetric_rate_smoothing(self):
        dates_fx_rates = [
            (datetime(2017, 2, 18), 2.9),
            (datetime(2017, 2, 19), np.nan),
            (datetime(2017, 2, 20), np.nan),
            (datetime(2017, 2, 21), np.nan),
            (datetime(2017, 2, 22), 3.1),
        ]
        converter = DummyRateConverterTo('FOO', dates_fx_rates)
        assert converter.get_rate('FOO', datetime(2017, 2, 20)) == 3.0


@pytest.fixture
def rate_converter():
    return DummyRateConverterTo('USD', [
        (datetime(2019, 4, 18), 1.1),
        (datetime(2017, 1, 20), 1.1),
        (datetime(2017, 2, 20), 1.1),
        (datetime(2017, 1, 1), 1.1),
        (datetime(2017, 2, 1), 1.1)
    ])


@pytest.fixture
def calculator(rate_converter):
    return expat_fatcat.FatcatCalculator(rate_converter)


@pytest.fixture
def salary():
    return [
         {'date': '2017-01-20', 'amount': 1000},
         {'date': '2017-02-20', 'amount': 1500}
    ]


class TestCalculator:
    """
    Tests for the main functionality of rate calculators.
    Excludes edge cases such as missing values, invalid arguments, etc.
    """

    def test_date_conversion(self, calculator):
        assert (
            calculator.parse_date('2019-04-18', '%Y-%m-%d')
            == datetime(2019, 4, 18)
        )

    def test_get_rate(self, calculator):
        rate = calculator.get_rate('FOO', '2019-04-18', '%Y-%m-%d')
        assert rate == 1.1

    def test_get_converted_amount(self, calculator):
        amount = 10.
        assert (calculator.get_converted_amount(
            amount, 'FOO', '2019-04-18', '%Y-%m-%d')
            == pytest.approx(11.)
        )

    def test_calculate_agg_payment(self, calculator, salary):
        agg_payment = calculator(salary, 'FOO', '%Y-%m-%d')
        assert agg_payment == pytest.approx(2750.)

    def test_convert_payments(self, calculator, salary):
        res = calculator.convert_payments(salary, 'FOO', '%Y-%m-%d')
        expected = [
            {'date': '2017-01-20', 'amount': 1000, 'converted_amount': 1100.},
            {'date': '2017-02-20', 'amount': 1500, 'converted_amount': 1650.}
        ]
        pd.testing.assert_frame_equal(
            pd.DataFrame(res),
            pd.DataFrame(expected)
        )

# # ######################################
# # # f2555
# # ######################################


@pytest.fixture
def rent():
    return [
         {'date': '2017-01-20', 'amount': 650},
         {'date': '2017-02-20', 'amount': 650},
     ]


@pytest.fixture
def dividends():
    return [
         {'date': '2017-01-01', 'amount': 10},
         {'date': '2017-02-01', 'amount': 50},
    ]


@pytest.fixture
def f2555_data(salary, rent, dividends):
    return [
        {'tag': 'rent', 'currency': 'FOO', 'payments': rent},
        {'tag': 'salary', 'currency': 'FOO', 'payments': salary},
        {'tag': 'dividends', 'currency': 'FOO', 'payments': dividends},
    ]


class TestF2555:
    """Tests for form 2555 class F2555"""
    def test_f2555_call(self, rate_converter, f2555_data):

        res = expat_fatcat.f2555(rate_converter, f2555_data, '%Y-%m-%d')
        assert res.get('rent').get('amount') == pytest.approx(1430.)
        assert res.get('rent').get('currency') == 'FOO'
        assert res.get('salary').get('amount') == pytest.approx(2750.)
        assert res.get('salary').get('currency') == 'FOO'
        assert res.get('dividends').get('amount') == pytest.approx(66.)
        assert res.get('dividends').get('currency') == 'FOO'


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'munichpavel/expat_fatcat'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


src/expat_fatcat/expat_fatcat.py
"""Main module."""
import os
import warnings
import numpy as np
import pandas as pd

from datetime import datetime, timedelta

from abc import ABC, abstractmethod

import quandl


class FatcatCalculator():
    """Calculate fx rates and other tax related quantities for IRS and FATCA
    compliance"""
    def __init__(self, rate_converter):
        self.rate_converter = rate_converter
        self._check_converter()

    def _check_converter(self):
        if self.rate_converter.to_currency != 'USD':
            raise NotImplementedError(
                'Invalid to_currency: only to USD implemented'
            )

    def convert_payments(self, payments, from_currency, date_format):
        """
        Convert payment history from input currency to rate_converter target
        currency.

        Parameters
        ----------
        from_currency : str
            Currency of payment
        payments : list of dicts

        Returns
        -------
        res : list of dicts
            Payments with extra field for fx converted amount
        """
        res = payments.copy()
        for payment in res:
            payment['converted_amount'] = self.get_converted_amount(
                payment.get('amount'), from_currency, payment.get('date'),
                date_format
            )

        return res

    def get_converted_amount(
        self, amount, from_currency,
        date_string, date_format
    ):
        return self.get_rate(from_currency, date_string, date_format) * amount

    def get_rate(self, from_currency, date_string, date_format):
        date = self.parse_date(date_string, date_format)
        rate = self.rate_converter.get_rate(from_currency, date)
        return rate

    def parse_date(self, date_string, date_format):
        return datetime.strptime(date_string, date_format)

    def __call__(self, payments, from_currency, date_format):

        res = self.convert_payments(payments, from_currency, date_format)
        return pd.DataFrame(res)['converted_amount'].sum()


class AbsRateConverterTo(ABC):
    """Abstract class for conversions to a given currency"""
    def __init__(self, to_currency):
        self.to_currency = to_currency

    def get_rate(self, from_currency, date, offset=0):
        """Returns exchange rate, with smoothing if no fx rate for date"""
        call_str = self._get_call_str(from_currency)
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", category=RuntimeWarning)
            res = np.nanmean([
                self._rate_api_call(call_str, date + timedelta(offset)),
                self._rate_api_call(call_str, date + timedelta(-offset))
            ])

        if not np.isnan(res):
            return res
        else:
            offset += 1
            return self.get_rate(from_currency, date, offset)

    @abstractmethod
    def _get_call_str(self, from_currency):
        pass

    @abstractmethod
    def _rate_api_call(self, call_str, date):
        pass


class QuandlRateConverterToUSD(AbsRateConverterTo):
    """Get conversion rates to USD from the QUANDL python api"""
    def __init__(self):
        self.to_currency = 'USD'
        self._set_key()

    def _set_key(self):
        if 'QUANDL_KEY' in os.environ:
            quandl.ApiConfig.api_key = os.environ['QUANDL_KEY']

    def _get_call_str(self, from_currency):
        if from_currency == 'EUR':
            return 'ECB/EURUSD'
        elif from_currency == 'GBP':
            return 'FRED/DEXUSUK'
        else:
            raise NotImplementedError(
                'Only EUR and GBP to USD rates are implemented'
            )

    def _rate_api_call(self, call_str, date):
        r = quandl.get(call_str, start_date=date, end_date=date)
        try:
            res = r.iloc[0]['Value']
        except IndexError:
            warnings.warn(
                "Date {} is invalid, trying before and after".format(date)
            )
            res = np.nan

        return res


class DummyRateConverterTo(AbsRateConverterTo):
    """Dummy fx rate converter for testing and development"""
    def __init__(self, to_currency, dates_fx_rates):
        self.to_currency = to_currency
        self._dates_fx_rates = dates_fx_rates

    def _get_call_str(self, from_currency):
        if from_currency == 'FOO':
            return 'ECB/FOO' + self.to_currency
        elif from_currency == 'BAR':
            return 'ECB/BAR' + self.to_currency
        else:
            raise NotImplementedError(
                'Only FOO or BAR to {} rates are implemented'
                .format(self.to_currency)
            )

    def _rate_api_call(self, call_str, date):
        """Returns dummy value for all argument values"""
        return self._lookup_rate(date)

    def _lookup_rate(self, date):
        for r in self._dates_fx_rates:
            if r[0] == date:
                return r[1]


def f2555(rate_converter, data, date_format):
    """
    Calculate total payments in USD for form f2555.

    Parameters
    ----------
    rate_converter : AbsRateConverterTo()
    data : list of dicts with fields tag, currency, payments
    date_format : date format string

    Example usage
    -------------
    >>> converter = DummyRateConverterTo('USD')
    >>> data = [
    ...     {'tag': 'salary', 'currency': 'FOO',
    ...      'payments':  [
    ...          {'date': '2017-01-20', 'amount': 1000},
    ...          {'date': '2017-02-20', 'amount': 1500}
    ...      ]},
    ...     {'tag': 'dividends', 'currency': 'BAR',
    ...      'payments': [
    ...          {'date': '2017-01-01', 'amount': 10},
    ...          {'date': '2017-02-01', 'amount': 50}
    ...      ]}
    ... ]
    >>> f2555(converter, data, '%Y-%m-%d')
    {'salary': {'currency': 'FOO', 'amount': 2812.5}, 'dividends':
    {'currency': 'BAR', 'amount': 67.5}}
    """
    calculator = FatcatCalculator(rate_converter)
    res = {}
    for d in data:
        res[d['tag']] = {
            'currency': d['currency'],
            'amount': calculator(d['payments'], d['currency'], date_format)
        }

    return res
